proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+    enableMechanism = ...
+        -- Internally this would be mutate program-global state
+
+    disableMechanism :: BacktraceMechanism -> IO ()
+    disableMechanism = ...
+
+A ``collectBacktrace`` primitive used by ``throwWithContext``
+simply dispatches to the currently-selected ``BacktraceMechanism``\ s: ::
+
+    module GHC.Exception.Backtrace where
+
+    -- | Collect an 'ExceptionContext' containing backtraces from all enabled
+    -- 'BacktraceMechanism's.
+    collectBacktraces :: HasCallStack => IO ExceptionContext
+    collectBacktraces = do
+        mechs <- readIORef currentBacktraceMechanisms
+        mconcat `fmap` mapM collectBacktrace mechs
+
+    -- | Collect a 'Backtrace' via the given 'BacktraceMechanism'.
+    collectBacktrace :: HasCallStack => BacktraceMechanism -> IO ExceptionContext
+
+
+    module GHC.Exception where
+
+    -- | Throw an exception. Exceptions may be thrown from purely
+    -- functional code, but may only be caught within the 'IO' monad.
+    -- Backtraces are collected using the backtrace mechanisms selected by
+    -- 'GHC.Exception.Backtrace.enabledBacktraceMechanisms'.
+    throwWithContext :: forall e a. (HasCallStack, Exception e)
+                     => e -> ExceptionContext -> a
+    throwWithContext e ctxt = do
+        -- (implementation simplified for clarity)
+        backtraces <- collectBacktraces
+        raise# (toExceptionWithContext (ctxt <> backtraces) e)
+
+Note that in order to provide ``HasCallStack`` backtraces we propose that a
+``HasCallStack`` constraint be added to ``throw``, ``throwIO``, and similar
+functions. Our prototype implementation suggests that this likely does not
+carry a significant performance impact.
+
+Since some users may want to explicitly opt out of backtrace collection when
+throwing certain exceptions (e.g. in codebases where exceptions are used for
+non-exceptional flow control), we also propose to add non-backtrace-collecting
+``throw`` variants: ::
+
+    throwNoBacktrace   :: forall e a. (Exception e) => e -> a
+    throwIONoBacktrace :: forall e a. (Exception e) => e -> a
+
+
+Teach top-level handler to use ``displayException``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For historical reasons, GHC's top-level exception handler currently displays
+unhandled exceptions using ``Show`` rather than the ``Exception`` class's
+``displayException``.
+
+Since only ``displayException`` will display exception
+context, we propose that this behavior is changed: unhandled exceptions
+should be displayed to the user using ``displayException``.
+As the default implementation of ``displayException`` simply delegates to
+``show``, we expect that the messages produced by most exceptions will be
+unaffected by this change (except for the context added by ``SomeException``\'s
+``displayException`` implementation).
+
+Examples
+--------
+
+User programs would typically call ``setEnabledBacktraceMechanisms`` during
```

Comment:
Looks like the current version of the proposal doesn't mention `setEnabledBacktraceMechanisms` in the API of `GHC.Exception.Backtrace`?